light-spike/src/components/demuxer.rs

```diff
@@ -0,0 +1,245 @@
+            std::thread::sleep(Duration::from_secs(1));
+        }
+    }
+
+    fn is_trusted(&self, light_block: &LightBlock) -> bool {
+        let in_store = self.state.trusted_store_reader.get(light_block.height);
+        in_store.as_ref() == Some(light_block)
+    }
+
+    pub fn sync_to_latest(&mut self) -> Result<(), Error> {
+        if self.state.trusted_store_reader.latest().is_none() {
+            bail!(ErrorKind::NoInitialTrustedState)
+        };
+
+        let primary = self.state.peers.primary.clone();
+        let target_block = match self.fetch_light_block(primary, LATEST_HEIGHT) {
+            Ok(last_block) => last_block,
+            Err(io_error) => bail!(ErrorKind::Io(io_error)),
+        };
+
+        if !self.is_trusted(&target_block) {
+            self.verify_to_target(target_block.height)?;
+        }
+
+        Ok(())
+    }
+
+    fn verify_to_target(&mut self, target_height: Height) -> Result<(), Error> {
+        let options = self.options.with_now(self.clock.now());
+
+        // TODO: Check this ahead of time
+        precondition!(
+            contracts::verify::trusted_state_contains_block_within_trusting_period(
+                &self.state.trusted_store_reader,
+                self.options.trusting_period,
+                options.now
+            )
+        );
+
+        // TODO: This might now be a good precondition if we need to verify
+        //       intermediate blocks, eg. for the relayer.
+        precondition!(
+            contracts::verify::target_height_greater_than_all_blocks_in_trusted_store(
+                target_height,
+                &self.state.trusted_store_reader,
+            )
+        );
+
+        let mut next_height = target_height;
+        let mut trusted_state = self.state.trusted_store_reader.latest().unwrap();
+
+        while trusted_state.height < target_height {
+            trusted_state = self.state.trusted_store_reader.latest().unwrap();
+
+            dbg!(target_height);
+            dbg!(trusted_state.height);
+            dbg!(next_height);
+
+            let primary = self.state.peers.primary.clone();
+            let current_block = match self.fetch_light_block(primary, next_height) {
+                Ok(current_block) => current_block,
+                Err(_) => return Ok(()),
+            };
+
+            let verification_result =
+                self.verify_light_block(&current_block, &trusted_state, &options);
+
+            dbg!(&verification_result);
+
+            if let VerifierOutput::Success = verification_result {
+                self.state.trusted_store_writer.add(current_block.clone());
+            } else {
+                self.state.untrusted_store_writer.add(current_block.clone());
+            }
+
+            self.state.trace_block(target_height, current_block.height);
+
+            let schedule = self.schedule(&current_block, &trusted_state, verification_result);
+            dbg!(&schedule);
+
+            match schedule {
```

Comment:
Here it seems like we are are sending invalid block to the scheduler just to them to come back as scheduler events. For instance `VerifierOutput::Invalid(e)` maps to `SchedulerOutput::InvalidLightBlock(e)` but really the validity of the block should be in the verifier domain and not the scheduler domain. Would it make sense in this case to just match the verifier output directly here? Something like:


```rust
 match verification_result {
    VerifierOutput::Success => {
        self.state.trusted_store_writer.add(current_block.clone());
        continue
    },
    VerifierOutput::Invalid(e) => {
        self.state.untrusted_store_writer.add(current_block.clone());
        bail!(ErrorKind::InvalidLightBlock(e));
    },
    VerifierOutput::NotEnoughTrust => {
        self.state.untrusted_store_writer.add(current_block.clone());
        let schedule = self.schedule(&current_block,
            &trusted_state,
            verification_result);

        dbg!(&schedule);

        match schedule {
            SchedulerOutput::NextHeight(height) if height <= trusted_state.height => {
                bail!(ErrorKind::BisectionFailed(target_height, height));
            }
            SchedulerOutput::NextHeight(height) => {
                postcondition!(contracts::schedule::postcondition(
                    &trusted_state,
                    target_height,
                    height,
                    &self.state.trusted_store_reader,
                    &self.state.untrusted_store_reader
                ));

                next_height = height;
            }
        }
    }
}
```

Maybe there is a bit more nesting but it keeps the logic somehow together and decouples the scheduler and verifier domains. Just an idea 🤔